File: src/main/java/org/starlisp/core/LispBignum.java
package org.starlisp.core;

import java.math.BigInteger;

final class LispBignum extends LispInteger {
  private BigInteger n;

  public static LispBignum parse(String str) {
    return new LispBignum(new BigInteger(str));
  }

  public LispBignum(BigInteger nbr) {
    n = nbr;
  }

  public LispBignum(long nbr) {
    n = BigInteger.valueOf(nbr);
  }

  public LispBignum(int nbr) {
    n = BigInteger.valueOf((long) nbr);
  }

  public int hashCode() {
    return n.hashCode();
  }

  public LispBignum add(LispBignum nbr) {
    return new LispBignum(n.add(nbr.n));
  }

  public LispBignum sub(LispBignum nbr) {
    return new LispBignum(n.subtract(nbr.n));
  }

  public LispBignum mul(LispBignum nbr) {
    return new LispBignum(n.multiply(nbr.n));
  }

  public LispNumber div(LispBignum nbr) {
    return new LispBignum(n.divide(nbr.n));
  } // TODO : RATIONALS

  public LispBignum mod(LispBignum nbr) {
    return new LispBignum(n.remainder(nbr.n));
  }

  public LispNumber add(LispNumber nbr) {
    return (nbr instanceof LispFlonum) ? (new LispFlonum(n.doubleValue())).add((LispFlonum) nbr) :
        (nbr instanceof LispFixnum) ? add(coerceFixnumToBignum(nbr)) :
            add((LispBignum) nbr);
  }

  public LispNumber sub(LispNumber nbr) {
    return (nbr instanceof LispFlonum) ? (new LispFlonum(n.doubleValue())).sub((LispFlonum) nbr) :
        (nbr instanceof LispFixnum) ? sub(coerceFixnumToBignum(nbr)) :
            sub((LispBignum) nbr);
  }

  public LispNumber mul(LispNumber nbr) {
    return (nbr instanceof LispFlonum) ? (new LispFlonum(n.doubleValue())).mul((LispFlonum) nbr) :
        (nbr instanceof LispFixnum) ? mul(coerceFixnumToBignum(nbr)) :
            mul((LispBignum) nbr);
  }

  public LispNumber div(LispNumber nbr) {
    return (nbr instanceof LispFlonum) ? (new LispFlonum(n.doubleValue())).div((LispFlonum) nbr) :
        (nbr instanceof LispFixnum) ? div(coerceFixnumToBignum(nbr)) :
            div((LispBignum) nbr);
  }

  public LispInteger mod(LispInteger nbr) {
    return (nbr instanceof LispFixnum) ? mod(coerceFixnumToBignum(nbr)) : mod((LispBignum) nbr);
  }

  public LispInteger ash(LispInteger nbr) {
    return new LispBignum(n.shiftLeft(nbr.toJavaInt()));
  } // TODO: Only well-defined between -2^31 and 2^31-1 inclusive

  public boolean equals(Object obj) {
    return (obj instanceof LispBignum) ? n.equals(((LispBignum) obj).n) :
        (obj instanceof LispFixnum) ? equals(coerceFixnumToBignum((LispFixnum) obj)) :
            (obj instanceof LispFlonum) ? obj.equals(this) :
                null;
  }

  public boolean negP() {
    return n.signum() == -1;
  }

  public String toString() {
    return n.toString();
  }

  public int toJavaInt() {
    return n.intValue();
  }

  public long toJavaLong() {
    return n.longValue();
  }

  public float toJavaFloat() {
    return n.floatValue();
  }

  public double toJavaDouble() {
    return n.doubleValue();
  }

  public BigInteger toJavaBigInteger() {
    return n;
  }
}

File: src/main/scala/org/starlisp/core/LispBignum.scala
package org.starlisp.core

import java.math.BigInteger

object LispBignum {
  def parse(str: String): LispBignum = new LispBignum(new BigInteger(str))
}

class LispBignum(val n: BigInteger) extends LispInteger {
  def this(nbr: Long) {
    this(BigInteger.valueOf(nbr))
  }

  def this(nbr: Int) {
    this(BigInteger.valueOf(nbr.asInstanceOf[Long]))
  }

  override def hashCode: Int = n.hashCode
  override def equals(obj: Any): Boolean = {
    if ((obj.isInstanceOf[LispBignum]))
      (n == (obj.asInstanceOf[LispBignum]).n)
    else if ((obj.isInstanceOf[LispFixnum]))
      (this == coerceFixnumToBignum(obj.asInstanceOf[LispFixnum]))
    else if ((obj.isInstanceOf[LispFlonum]))
      (obj == this)
    else
      false
  }

  def add(nbr: LispBignum): LispBignum = new LispBignum(n.add(nbr.n))
  def sub(nbr: LispBignum): LispBignum = new LispBignum(n.subtract(nbr.n))
  def mul(nbr: LispBignum): LispBignum = new LispBignum(n.multiply(nbr.n))
  def div(nbr: LispBignum): LispNumber = new LispBignum(n.divide(nbr.n))
  def mod(nbr: LispBignum): LispBignum = new LispBignum(n.remainder(nbr.n))
  def ash(nbr: LispInteger): LispInteger = new LispBignum(n.shiftLeft(nbr.toJavaInt))
  def negP: Boolean = n.signum == -1

  def add(nbr: LispNumber): LispNumber = {
    if ((nbr.isInstanceOf[LispFlonum])) (new LispFlonum(n.doubleValue)).add(nbr.asInstanceOf[LispFlonum]) else if ((nbr.isInstanceOf[LispFixnum])) add(coerceFixnumToBignum(nbr)) else add(nbr.asInstanceOf[LispBignum])
  }

  def sub(nbr: LispNumber): LispNumber = {
    if ((nbr.isInstanceOf[LispFlonum])) (new LispFlonum(n.doubleValue)).sub(nbr.asInstanceOf[LispFlonum]) else if ((nbr.isInstanceOf[LispFixnum])) sub(coerceFixnumToBignum(nbr)) else sub(nbr.asInstanceOf[LispBignum])
  }

  def mul(nbr: LispNumber): LispNumber = {
    if ((nbr.isInstanceOf[LispFlonum])) (new LispFlonum(n.doubleValue)).mul(nbr.asInstanceOf[LispFlonum]) else if ((nbr.isInstanceOf[LispFixnum])) mul(coerceFixnumToBignum(nbr)) else mul(nbr.asInstanceOf[LispBignum])
  }

  def div(nbr: LispNumber): LispNumber = {
    if ((nbr.isInstanceOf[LispFlonum])) (new LispFlonum(n.doubleValue)).div(nbr.asInstanceOf[LispFlonum]) else if ((nbr.isInstanceOf[LispFixnum])) div(coerceFixnumToBignum(nbr)) else div(nbr.asInstanceOf[LispBignum])
  }

  def mod(nbr: LispInteger): LispInteger = {
    if ((nbr.isInstanceOf[LispFixnum])) mod(coerceFixnumToBignum(nbr)) else mod(nbr.asInstanceOf[LispBignum])
  }

  override def toString: String = n.toString

  def toJavaInt: Int = n.intValue
  def toJavaLong: Long = n.longValue
  def toJavaFloat: Float = n.floatValue
  def toJavaDouble: Double = n.doubleValue
  def toJavaBigInteger: BigInteger = n
}